core/src/main/java/org/apache/carbondata/core/memory/UnsafeMemoryManager.java

```diff
@@ -105,13 +105,15 @@ private synchronized MemoryBlock allocateMemory(long taskId, long memoryRequeste
   }
 
   public synchronized void freeMemory(long taskId, MemoryBlock memoryBlock) {
-    taskIdToMemoryBlockMap.get(taskId).remove(memoryBlock);
-    allocator.free(memoryBlock);
-    memoryUsed -= memoryBlock.size();
-    memoryUsed = memoryUsed < 0 ? 0 : memoryUsed;
-    LOGGER.info(
-        "Freeing memory of size: " + memoryBlock.size() + "available memory:  " + (totalMemory
-            - memoryUsed));
+    if (taskIdToMemoryBlockMap.containsKey(taskId)) {
```

Comment:
we cannot just rely on the `taskIdToMemoryBlockMap` , if code requests to clean memoryblock then we must clean it. so `allocator.free(memoryBlock)` has to be called in all scenarios. So introduce a boolean flag inside `MemoryBlock` to check already cleaned or not . if it is not cleaned then clean the block the set the boolean to `true`. Same needs to be added in `freeMemoryAll` also.